Aws/S3/Core.hs

```diff
@@ -240,6 +276,178 @@ s3SignQuery S3Query{..} S3Configuration{..} SignatureData{..}
+            , mconcat . catMaybes $ path             -- path
+            , s3RenderQuery False $ sort queryString -- query string
+            ] ++
+            Map.foldMapWithKey (\a b -> [CI.foldedCase a <> ":" <> b]) canonicalHeaders ++
+            [ "" -- end headers
+            , signedHeaders
+            , amzHeaders Map.! hAmzContentSha256
+            ]
+
+        (authorization, signatureQuery, queryString) = case ti of
+            AbsoluteTimestamp _  -> (Just . return $ auth, [], allQueries)
+            AbsoluteExpires time ->
+                ( Nothing
+                , [(CI.original hAmzSignature, Just sig)]
+                , (allQueries ++) . HTTP.toQuery . map (first CI.original) $
+                    [ (hAmzAlgorithm, "AWS4-HMAC-SHA256")
+                    , (hAmzCredential, cred)
+                    , (hAmzDate, amzHeaders Map.! hAmzDate)
+                    , (hAmzExpires, B8.pack . (show :: Integer -> String) . floor $ diffUTCTime time signatureTime)
+                    , (hAmzSignedHeaders, signedHeaders)
+                    ] ++ iamTok
+                )
+            where
+                allQueries = s3QSubresources ++ s3QQuery
+                region = s3ExtractRegion s3Endpoint
+                auth = authorizationV4' sd HmacSHA256 region "s3" signedHeaders stringToSign
+                sig  = signatureV4      sd HmacSHA256 region "s3"               stringToSign
+                cred = credentialV4     sd            region "s3"
+                ti = case (s3UseUri, signatureTimeInfo) of
+                    (False, t) -> t
+                    (True, AbsoluteTimestamp time) -> AbsoluteExpires $ s3DefaultExpiry `addUTCTime` time
+                    (True, AbsoluteExpires time) -> AbsoluteExpires time
+
+-- | Custom UriEncode function
+-- see <http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html>
+s3UriEncode
+    :: Bool         -- ^ Whether encode slash characters
+    -> B.ByteString
+    -> B.ByteString
+s3UriEncode encodeSlash = B8.concatMap $ \c ->
+    if (isAscii c && isAlphaNum c) || (c `elem` nonEncodeMarks)
+        then B8.singleton c
+        else B8.pack $ '%' : map toUpper (showHex (ord c) "")
+    where
+        nonEncodeMarks :: String
+        nonEncodeMarks = if encodeSlash
+            then "_-~."
+            else "_-~./"
+
+s3RenderQuery
+    :: Bool -- ^ Whether prepend a question mark
+    -> HTTP.Query
+    -> B.ByteString
+s3RenderQuery qm = mconcat . qmf . intersperse (B8.singleton '&') . map renderItem
+    where
+        qmf = if qm then ("?":) else id
+
+        renderItem :: HTTP.QueryItem -> B8.ByteString
+        renderItem (k, Just v) = s3UriEncode True k <> "=" <> s3UriEncode True v
+        renderItem (k, Nothing) = s3UriEncode True k <> "="
+
+-- | see: <http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region>
+s3ExtractRegion :: B.ByteString -> B.ByteString
+s3ExtractRegion "s3.us-east-2.amazonaws.com"           = "us-east-2"
```

Comment:
That seems cumbersome if a new region is added

Reply:
I think it is too.
But there are breaking changes if we replace `s3Endpoint` with `s3Region` (#199's approach).
May we go with the approach?

Another idea is parsing the domain, but there is a similar issue yet.
I think the parser should be fixed if a new domain/service is added.

Reply:
Anyway, I have tried using parsing approach.
I think there is something better than previous one.